# **Dataset Prediksi Kesehatan Mental Siswa**

Data yang digunakan termuat dalam kaggel dengan url sebagai berikut :

https://www.kaggle.com/datasets/utkarshsharma11r/

## 📥 Data Loading

In [2]:
import pandas as pd
student_mental_health = pd.read_csv('/content/StudentMentalHealth_dirty.csv')

In [ ]:
# Eksekusi kedua sintaks
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.float_format', '{:.2f}'.format)

# **Identifikasi Masalah Utama**

Peralihan pembelajaran pada saat COVID-19 mengubah pola belajar, interaksi sosial, dan kebiasaan harian siswa, yang berpotensi memengaruhi kesehatan mental mereka, maka dilakukan analisis untuk mengetahui pola hubungan antarvariabel, sehingga informasi ini dapat menjadi dasar bagi pendidik dan pemangku kebijakan dalam merancang intervensi atau dukungan yang lebih tepat sasaran.

## 🎯 Tujuan Project
Project ini bertujuan untuk:
- Menganalisis faktor-faktor yang memengaruhi kesehatan mental mahasiswa
- Mengidentifikasi pola antara screen time, durasi tidur, aktivitas fisik, dan level stres
- Memberikan insight berbasis data yang relevan untuk lingkungan akademik

# **Informasi Umum Pada Data**

In [ ]:
display(student_mental_health)

Name  Gender   Age Education Level Screen Time (hrs/day)  \
0       Aarav    Male  15.0         Class 8                   7.1   
1       Meera  Female  25.0             MSc                   3.3   
2      Ishaan    Male  20.0           BTech                   9.5   
3      Aditya    Male  20.0              BA                  10.8   
4       Anika  Female  17.0        Class 11                   2.8   
...       ...     ...   ...             ...                   ...   
1245   Vivaan    Male  21.0           BTech                   6.1   
1246    Arjun    Male  15.0         Class 9                   5.5   
1247    Aarav    Male  22.0             NaN                   5.5   
1248   Ananya  Female  16.0        Class 10                  10.4   
1249  Krishna    Male  24.0           MTech                   2.0   

      Sleep Duration (hrs)  Physical Activity (hrs/week) Stress Level  \
0                     8.90                          9.30       Medium   
1                    -5.00                          0.20       Medium   
2                   150.00                          6.20       Medium   
3                     5.60                          5.50         High   
4                     5.40                          3.10       Medium   
...                    ...                           ...          ...   
1245                  6.20                          4.00          Low   
1246                  8.00                          6.90       Medium   
1247                  4.40                         -5.00       Medium   
1248                  8.00                          6.60         High   
1249                  8.00                          8.10         High   

     Anxious Before Exams Academic Performance Change  
0                      No                        Same  
1                      No                        Same  
2                      No                        Same  
3                     Yes                        Same  
4                     Yes                        Same  
...                   ...                         ...  
1245                  Yes                    Declined  
1246                  Yes                        Same  
1247                   No                    Declined  
1248                   No                    Declined  
1249                  Yes                        Same  

[1250 rows x 10 columns]

In [ ]:
print(student_mental_health.columns)

Index(['Name', 'Gender', 'Age', 'Education Level', 'Screen Time (hrs/day)',
       'Sleep Duration (hrs)', 'Physical Activity (hrs/week)', 'Stress Level',
       'Anxious Before Exams', 'Academic Performance Change'],
      dtype='object')


In [ ]:
print(student_mental_health.index)

RangeIndex(start=0, stop=1250, step=1)


## 🧠 Executive Summary

Dari analisis terhadap **1251 data mahasiswa**, ditemukan bahwa:

- Mayoritas mahasiswa mengalami **tingkat stres sedang hingga tinggi**
- Mahasiswa dengan **durasi tidur rendah (<7 jam)** cenderung memiliki stres lebih tinggi
- **Screen time berlebih (>6 jam/hari)** berasosiasi dengan peningkatan kecemasan sebelum ujian
- Aktivitas fisik rutin berkorelasi dengan **penurunan tingkat stres**

📌 Temuan ini menunjukkan pentingnya keseimbangan gaya hidup digital, tidur, dan aktivitas fisik dalam menjaga kesehatan mental mahasiswa.

In [ ]:
student_mental_health.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1250 entries, 0 to 1249
Data columns (total 10 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Name                          1240 non-null   object 
 1   Gender                        1239 non-null   object 
 2   Age                           1240 non-null   object 
 3   Education Level               1240 non-null   object 
 4   Screen Time (hrs/day)         1236 non-null   object 
 5   Sleep Duration (hrs)          1232 non-null   float64
 6   Physical Activity (hrs/week)  1243 non-null   float64
 7   Stress Level                  1238 non-null   object 
 8   Anxious Before Exams          1233 non-null   object 
 9   Academic Performance Change   1235 non-null   object 
dtypes: float64(2), object(8)
memory usage: 97.8+ KB


| Kolom | Deskripsi | Satuan |
|------|----------|------|
| Name | Nama siswa yang menjadi responden | - |
| Gender | Jenis kelamin siswa | - |
| Age | Usia siswa | Tahun |
| Education level | Tingkat pendidikan yang sedang ditempuh | - |
| Screen Time | Durasi penggunaan layar | Jam/hari |
| Sleep Duration | Durasi tidur |  Jam |
| Physical Activity | Durasi aktivitas fisik | Jam/minggu |
| Stress Level | Tingkat stres | Skala |
| Anxious Before Exams | Tingkat kecemasan sebelum ujian | Ya/Tidak |
| Academic Performance Change | Perubahan akademik siswa | - |

In [ ]:
pip install ydata-profiling

In [ ]:
import ydata_profiling as yp

# Dapatkan summary report
profile = yp.ProfileReport(
    student_mental_health,
    title = 'Laporan Profiling student mental health'
)

# Simpan hasilnya ke format html
profile.to_file('profiling_report_student_mental_health.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|██████████| 10/10 [00:00<00:00, 24.06it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

# 🧹 **DATA CLEANING**

In [ ]:
# mengurutkan berdasarkan umur secara ascending
student_mental_health = student_mental_health.sort_values(by = ['Age'], ascending = True)
display(student_mental_health)

Name  Gender  Age Education Level Screen Time (hrs/day)  \
1092     Diya  Female  1.0           BTech                   7.2   
1087    Meera  Female  1.0              MA                   5.3   
555   Shaurya    Male  1.0         Class 9                   9.8   
1061      Sai    Male  1.0           BTech                  11.2   
292    Saanvi  Female  1.0        Class 11                   2.0   
...       ...     ...  ...             ...                   ...   
522      Diya  Female  NaN           MTech                   9.4   
619    Aditya    Male  NaN           MTech                 150.0   
812     Kiran   Other  NaN             BSc                   2.5   
1083    Arjun    Male  NaN              MA                   5.4   
1157    Meera  Female  NaN             MSc                   9.6   

      Sleep Duration (hrs)  Physical Activity (hrs/week) Stress Level  \
1092                  6.40                          7.20       Medium   
1087                  6.50                          1.90       Medium   
555                  -5.00                          7.30       Medium   
1061                  5.40                          3.60       Medium   
292                   4.10                          0.10          Low   
...                    ...                           ...          ...   
522                   4.00                          9.00          Low   
619                   6.80                          1.90       Medium   
812                   6.60                          3.70         High   
1083                  5.10                          7.00       Medium   
1157                  7.20                          1.10          NaN   

     Anxious Before Exams Academic Performance Change  
1092                   No                    Improved  
1087                  Yes                        Same  
555                    No                    Improved  
1061                  Yes                    Improved  
292                   Yes                    Declined  
...                   ...                         ...  
522                    No                    Improved  
619                    No                        Same  
812                    No                        Same  
1083                  Yes                    Declined  
1157                  Yes                    Improved  

[1250 rows x 10 columns]

In [ ]:
# mengurutkan berdasarkan index
student_mental_health = student_mental_health.sort_values(by=['Age'], ascending=True, ignore_index=True)
display(student_mental_health)

Name  Gender  Age Education Level Screen Time (hrs/day)  \
0        Diya  Female  1.0           BTech                   7.2   
1         Sai    Male  1.0        Class 11                   2.2   
2      Saanvi  Female  1.0              MA                  10.7   
3     Krishna    Male  1.0           BTech                   6.9   
4       Anika  Female  1.0           BTech                   2.4   
...       ...     ...  ...             ...                   ...   
1245     Diya  Female  NaN           MTech                   9.4   
1246   Aditya    Male  NaN           MTech                 150.0   
1247    Kiran   Other  NaN             BSc                   2.5   
1248    Arjun    Male  NaN              MA                   5.4   
1249    Meera  Female  NaN             MSc                   9.6   

      Sleep Duration (hrs)  Physical Activity (hrs/week) Stress Level  \
0                     6.40                          7.20       Medium   
1                     4.20                          6.90       Medium   
2                     4.30                          5.00       Medium   
3                     4.20                          8.50          Low   
4                      NaN                          9.00       Medium   
...                    ...                           ...          ...   
1245                  4.00                          9.00          Low   
1246                  6.80                          1.90       Medium   
1247                  6.60                          3.70         High   
1248                  5.10                          7.00       Medium   
1249                  7.20                          1.10          NaN   

     Anxious Before Exams Academic Performance Change  
0                      No                    Improved  
1                      No                    Declined  
2                      No                    Improved  
3                      No                    Declined  
4                      No                        Same  
...                   ...                         ...  
1245                   No                    Improved  
1246                   No                        Same  
1247                   No                        Same  
1248                  Yes                    Declined  
1249                  Yes                    Improved  

[1250 rows x 10 columns]

In [ ]:
# Transformasi tipe data

# Mengonversi 'Age' ke numerik terlebih dahulu (menangani error dengan 'coerce'), lalu ke Int64Dtype
student_mental_health['Age'] = pd.to_numeric(student_mental_health['Age'], errors='coerce')
student_mental_health['Age'] = student_mental_health['Age'].astype(pd.Int64Dtype())

# Mengonversi 'Screen Time (hrs/day)' ke numerik (menangani error dengan 'coerce'), lalu ke Float64Dtype
student_mental_health['Screen Time (hrs/day)'] = pd.to_numeric(student_mental_health['Screen Time (hrs/day)'], errors='coerce')
student_mental_health['Screen Time (hrs/day)'] = student_mental_health['Screen Time (hrs/day)'].astype(pd.Float64Dtype())

# Tampilkan hasilnya
display(student_mental_health.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1250 entries, 0 to 1249
Data columns (total 10 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Name                          1240 non-null   object 
 1   Gender                        1239 non-null   object 
 2   Age                           1239 non-null   Int64  
 3   Education Level               1240 non-null   object 
 4   Screen Time (hrs/day)         1235 non-null   Float64
 5   Sleep Duration (hrs)          1232 non-null   float64
 6   Physical Activity (hrs/week)  1243 non-null   float64
 7   Stress Level                  1238 non-null   object 
 8   Anxious Before Exams          1233 non-null   object 
 9   Academic Performance Change   1235 non-null   object 
dtypes: Float64(1), Int64(1), float64(2), object(6)
memory usage: 100.2+ KB


None

In [ ]:
# mendefinisikan kolom numerik
kolom_numerik = student_mental_health.select_dtypes(include = ['number']).columns

# Tampilkan hasilnya
print(kolom_numerik)

Index(['Age', 'Screen Time (hrs/day)', 'Sleep Duration (hrs)',
       'Physical Activity (hrs/week)'],
      dtype='object')


In [ ]:
# mendefinisikan kolom kategorik
kolom_kategorik = student_mental_health.select_dtypes(exclude = ['number']).columns

# Tampilkan hasilnya
print(kolom_kategorik)

Index(['Name', 'Gender', 'Education Level', 'Stress Level',
       'Anxious Before Exams', 'Academic Performance Change'],
      dtype='object')


## 💡 **Insight**

Pada tahap ini dilakukan transformasi tipe data pada variabel **Age menjadi integer** dan **Screen Time menjadi float** guna memastikan kesesuaian tipe data numerik serta mendukung proses analisis dan visualisasi data secara optimal.



# **Missing Value Numerik**

In [ ]:
# mengisi missing value dengan median
imputasi_median = ['Age', 'Screen Time (hrs/day)', 'Sleep Duration (hrs)', 'Physical Activity (hrs/week)']

for col in imputasi_median:
    if student_mental_health[col].isnull().sum() > 0:
        median_value = student_mental_health[col].median()
        student_mental_health[col] = student_mental_health[col].fillna(median_value)
        print(f"Missing values di kolom '{col}' diisi dengan median: {median_value:.2f}")

In [ ]:
# Tampilkan missing value pada data numerik
display(student_mental_health[kolom_numerik].isna().sum())

Age                             0
Screen Time (hrs/day)           0
Sleep Duration (hrs)            0
Physical Activity (hrs/week)    0
dtype: int64

# 💡 **Insight**
* **Age** disi dengan
* **Screen Time (hrs/day)** diisi dengan median: 6.90
* **Sleep Duration (hrs)** diisi dengan median: 6.50
* **Physical Activity (hrs/week)** diisi dengan median: 5.10

# **Missing Value Kategorik**

In [ ]:
# mengisi missing value dengan modus
for col in kolom_kategorik:
    if student_mental_health[col].isnull().sum() > 0:
        mode_value = student_mental_health[col].mode()[0] # .mode() bisa mengembalikan beberapa jika ada banyak mode, kita ambil yang pertama
        student_mental_health[col] = student_mental_health[col].fillna(mode_value)
        print(f"Missing values di kolom '{col}' diisi dengan mode: '{mode_value}'")

In [ ]:
# Tampilkan missing value pada data kategorik

display(student_mental_health[kolom_kategorik].isna().sum())

Name                           0
Gender                         0
Education Level                0
Stress Level                   0
Anxious Before Exams           0
Academic Performance Change    0
dtype: int64

# 💡 **Insight**
* **Name** Diisi dengan 'Shaurya'
* **Gender** Diisi dengan modus 'Female'
* **Education Level** Diisi dengan modus 'MSc'
* **Stress Level** Diisi dengan modus 'Medium'
* **Anxious Before Exams** Diisi dengan modus 'Yes'
* **Academic Performance Change** Diisi dengan modus 'Same'

# **Redudansi Data**



In [ ]:
# Periksa duplikasi kolom

# Buat kondisi
conditions = student_mental_health.duplicated(subset = ['Name', 'Gender', 'Age', 'Education Level', 'Screen Time (hrs/day)',
       'Sleep Duration (hrs)', 'Physical Activity (hrs/week)', 'Stress Level',
       'Anxious Before Exams', 'Academic Performance Change'], keep = False)

# Tampilkan hasilnya
display(student_mental_health[conditions])

Name  Gender  Age Education Level  Screen Time (hrs/day)  \
13       Riya  Female   15         Class 8                   3.10   
16      Arjun    Male   15         Class 9                   5.50   
17    Shaurya    Male   15         Class 8                   5.00   
25        Dev   Other   15         Class 9                   7.80   
44      Kavya  Female   15         Class 8                  11.00   
...       ...     ...  ...             ...                    ...   
1215  Shaurya    Male   26             MSc                   3.80   
1218     Myra  Female   26             MSc                   6.10   
1219    Arjun    Male   26             MSc                   3.60   
1220     Myra  Female   26           MTech                   4.90   
1230     Myra  Female   26             MSc                   6.10   

      Sleep Duration (hrs)  Physical Activity (hrs/week) Stress Level  \
13                    7.60                          5.80          Low   
16                    8.00                          6.90       Medium   
17                    6.40                          5.50         High   
25                    4.20                          4.60         High   
44                    8.50                          1.90       Medium   
...                    ...                           ...          ...   
1215                  5.90                          2.80       Medium   
1218                  5.30                          9.30       Medium   
1219                  8.00                          5.80         High   
1220                  7.70                          8.90       Medium   
1230                  5.30                          9.30       Medium   

     Anxious Before Exams Academic Performance Change  
13                     No                    Improved  
16                    Yes                        Same  
17                    Yes                    Declined  
25                    Yes                    Improved  
44                     No                    Improved  
...                   ...                         ...  
1215                   No                        Same  
1218                  Yes                        Same  
1219                   No                    Improved  
1220                  Yes                    Improved  
1230                  Yes                        Same  

[355 rows x 10 columns]

In [ ]:
# Periksa ukuran data sebelum di drop duplicates
display(student_mental_health.shape)

(1250, 10)

In [ ]:
# Hapus data duplikat
student_mental_health = student_mental_health.drop_duplicates(subset = ['Name', 'Gender', 'Age', 'Education Level', 'Screen Time (hrs/day)',
       'Sleep Duration (hrs)', 'Physical Activity (hrs/week)', 'Stress Level',
       'Anxious Before Exams', 'Academic Performance Change'], keep = 'last')

# Periksa ukuran data setelah di drop duplicates
display(student_mental_health.shape)

(1064, 10)

 # 💡 **Insight**
Dataset awal mengandung sejumlah data duplikat. Setelah dilakukan proses penghapusan duplikasi berdasarkan variabel utama, jumlah data berkurang dari 1250 menjadi 1064 baris. Hal ini menunjukkan bahwa sekitar 15% data awal merupakan data redundan yang perlu dihapus untuk meningkatkan kualitas analisis.

In [ ]:
# Statistik Deskriptif Numerik

display(student_mental_health[kolom_numerik].describe().T)

count  mean   std   min   25%   50%   75%  \
Age                          1239.00 21.53 15.44  1.00 17.00 20.00 23.00   
Screen Time (hrs/day)        1235.00  8.17 13.47 -5.00  4.40  6.90  9.50   
Sleep Duration (hrs)         1232.00  8.20 15.18 -5.00  5.10  6.50  7.80   
Physical Activity (hrs/week) 1243.00  7.03 16.19 -5.00  2.70  5.10  7.70   

                                max  
Age                          200.00  
Screen Time (hrs/day)        150.00  
Sleep Duration (hrs)         150.00  
Physical Activity (hrs/week) 150.00

 # 💡 **Insight**
*   Mayoritas mahasiswa menghabiskan waktu lebih banyak di depan layar ($6.9$ jam) daripada untuk tidur ($6.5$ jam).
*   Indikasi adanya nilai mustahil (usia 200 tahun, waktu tidur -5 jam) menunjukkan data butuh pembersihan (cleaning).



In [ ]:
# Statistik Deskriptif kolom Kategorik

# Seleksi kolom dengan tipe object
kolom_kategorik = student_mental_health.columns

# Tampilkan frekuensinya
for col in kolom_kategorik:
    print(f'> Frekuensi \033[91m{col}\033[0m')
    print(f'  Terdapat {student_mental_health[col].nunique(dropna = False)} data unik\n')
    display(student_mental_health[col].value_counts(dropna = False).reset_index())
    print('\n')

> Frekuensi Name
  Terdapat 31 data unik



Name  count
0   Shaurya     77
1     Kavya     68
2     Meera     66
3    Aadhya     66
4      Diya     64
5     Arjun     64
6     Anika     61
7   Krishna     61
8      Myra     60
9   Reyansh     60
10    Aarav     60
11   Aditya     59
12   Ishaan     56
13   Ananya     54
14     Riya     54
15      Sai     51
16      Ira     50
17   Saanvi     49
18   Vivaan     48
19   Vihaan     41
20      Sam     16
21    Kiran     11
22      NaN     10
23    Rishi      8
24     Yash      7
25     Tanu      7
26      Dee      7
27      Dev      6
28     Tara      3
29     Arya      3
30      Ish      3



> Frekuensi Gender
  Terdapat 4 data unik



Gender  count
0  Female    592
1    Male    578
2   Other     69
3     NaN     11



> Frekuensi Age
  Terdapat 16 data unik



Age  count
0     17    121
1     21    120
2     23    115
3     15    108
4     20    104
5     16    104
6     19     96
7     25     96
8     26     90
9     22     88
10    18     88
11    24     85
12     1     13
13  <NA>     11
14   200      6
15   150      5



> Frekuensi Education Level
  Terdapat 12 data unik



Education Level  count
0              MSc    172
1            MTech    172
2               MA    164
3         Class 10    116
4         Class 11    115
5              BSc    105
6            BTech    104
7          Class 9    100
8               BA     78
9          Class 8     58
10        Class 12     56
11             NaN     10



> Frekuensi Screen Time (hrs/day)
  Terdapat 105 data unik



Screen Time (hrs/day)  count
0                      6.9     22
1                      4.8     21
2                      4.5     21
3                      6.3     20
4                      9.9     19
..                     ...    ...
100                   12.0      6
101                    6.2      6
102                    5.0      6
103                    9.2      4
104                    8.5      4

[105 rows x 2 columns]



> Frekuensi Sleep Duration (hrs)
  Terdapat 55 data unik



Sleep Duration (hrs)  count
0                    8.0     41
1                    8.2     34
2                    8.5     33
3                    5.6     33
4                    4.6     32
5                    4.2     32
6                    4.3     31
7                    4.8     31
8                    4.1     30
9                    6.8     30
10                   6.4     29
11                   8.9     28
12                   7.1     28
13                   7.2     27
14                   6.6     26
15                   4.4     26
16                   7.9     25
17                   6.7     25
18                   5.1     25
19                   6.1     24
20                   4.7     24
21                   5.4     24
22                   7.5     24
23                   6.0     24
24                   8.3     23
25                   6.5     23
26                   7.8     22
27                   7.6     22
28                   5.3     22
29                   8.8     22
30                   4.9     22
31                   5.8     22
32                   7.4     22
33                   7.0     21
34                   5.5     21
35                   7.3     21
36                   7.7     21
37                   5.0     20
38                   6.3     19
39                   6.2     19
40                   6.9     19
41                   8.7     19
42                   NaN     18
43                   5.9     18
44                   5.2     18
45                   8.6     16
46                   5.7     16
47                   9.0     14
48                   4.5     14
49                   8.4     14
50                   8.1     14
51                  -5.0     12
52                   4.0     11
53                 150.0     10
54                  99.0      9



> Frekuensi Physical Activity (hrs/week)
  Terdapat 105 data unik



Physical Activity (hrs/week)  count
0                             5.8     29
1                             3.2     26
2                             9.9     21
3                             1.9     20
4                             6.8     19
..                            ...    ...
100                           4.1      5
101                          10.0      5
102                           7.1      5
103                           1.3      5
104                           7.5      4

[105 rows x 2 columns]



> Frekuensi Stress Level
  Terdapat 4 data unik



Stress Level  count
0       Medium    625
1          Low    398
2         High    215
3          NaN     12



> Frekuensi Anxious Before Exams
  Terdapat 3 data unik



Anxious Before Exams  count
0                  Yes    634
1                   No    599
2                  NaN     17



> Frekuensi Academic Performance Change
  Terdapat 4 data unik



Academic Performance Change  count
0                        Same    486
1                    Improved    377
2                    Declined    372
3                         NaN     15

#  📈  **Visualisasi Data**

In [ ]:
import plotly.express as px

# Mendefinisikan kolom_numerik
kolom_numerik = student_mental_health.select_dtypes(include = ['number']).columns

color_map = {}
default_colors = ['#4E79A7', '#F28E2B', '#E15759', '#76B7B2', '#59A14F', '#EDC948', '#B07AA1']
for i, col_name in enumerate(kolom_numerik):
    color_map[col_name] = default_colors[i % len(default_colors)]

# Plot histogram masing-masing kolom numerik
for col in kolom_numerik:
    fig = px.histogram(
        student_mental_health,
        x = col,
        # nbins = 50,
        color_discrete_sequence = [color_map.get(col)],
        marginal = "box",
        hover_data=student_mental_health.columns
    )

    fig.update_yaxes(
        showgrid = False,
        showticklabels=False,
        title =''
    )

    fig.update_layout(
        title={
          'text' : f'Distribusi <b><span style="color:{color_map.get(col)}">{col}</span></b> Siswa<br><sup>Kesehatan Mental Siswa</sup>',
            'y':0.92,
            'x':0.5,
            'xanchor': 'center',
            'yanchor': 'top'
        },
        plot_bgcolor = 'rgba(0,0,0,0)',
        bargap = 0.01,
        title_font = dict(size = 25)
    )

    fig.show()

In [ ]:
import plotly.express as px

def box_plot(series, column_name, color):
    # Buat horizontal box plot
    fig = px.box(
        series,
        orientation = 'h',
        color_discrete_sequence  = [color]
    )

    # Update layout and display the plot
    fig.update_layout(
        title = f'<b>Box Plot {column_name}</b>',
        yaxis = dict(
            title = '',
            showgrid = False,
            showline = False,
            showticklabels = False,
            zeroline = False,
        ),
        xaxis = dict(
            title = column_name,
            showgrid = False,
            showline = True,
            showticklabels = True,
            zeroline = False,
        )
    )

    fig.show()

In [ ]:
student_mental_health.columns

Index(['Name', 'Gender', 'Age', 'Education Level', 'Screen Time (hrs/day)',
       'Sleep Duration (hrs)', 'Physical Activity (hrs/week)', 'Stress Level',
       'Anxious Before Exams', 'Academic Performance Change'],
      dtype='object')

In [ ]:
# Lakukan mapping kolom beserta warna
color_map = {
  'Age' : '#4E79A7',
  'Screen Time (hrs/day)' : '#F28E2B',
  'Sleep Duration (hrs)' : '#E15759',
  'Physical Activity (hrs/week)' : '#76B7B2'
}

# Terapkan ke semua kolom
for column, color in color_map.items():
    box_plot(student_mental_health[column], column, color)

# **Outlier**

In [ ]:
import plotly.express as px

def box_plot(series, column_name, color):
    # Buat horizontal box plot
    fig = px.box(
        series,
        orientation = 'h',
        color_discrete_sequence  = [color]
    )

    # Membuat layout dan display
    fig.update_layout(
        title = f'<b>Box Plot {column_name}</b>',
        yaxis = dict(
            title = '',
            showgrid = False,
            showline = False,
            showticklabels = False,
            zeroline = False,
        ),
        xaxis = dict(
            title = column_name,
            showgrid = False,
            showline = True,
            showticklabels = True,
            zeroline = False,
        )
    )

    fig.show()

In [ ]:
for col in kolom_numerik:
    box_plot(student_mental_health[col], col, '#B07AA1')

# 💡 **Insight**
Terdapat beberapa nilai ekstrem (outlier) pada variabel numerik, sebagian besar diduga akibat kesalahan input

### **Gunakan Teknik Winsorizing untuk mengatasi Outlier**
Teknik Winsorizing digunakan untuk mengatasi outlier dengan membatasi nilai ekstrem pada batas persentil tertentu. Pendekatan ini bertujuan untuk mengurangi pengaruh nilai ekstrem terhadap hasil analisis tanpa menghilangkan data, sehingga distribusi data menjadi lebih stabil dan representatif.

In [ ]:
# Fungsi untuk teknik winsorizing
def teknik_winsorizing(series):
    # Hitung Q1, Q3, dan IQR
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1

    # Hitung lower dan upper bound
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Jika lower bound negatif maka ubah menjadi 0
    if(lower_bound < 0):
      lower_bound = 0

    # Winsorizing: clip nilai ke batas bawah & atas
    series = series.astype(pd.Float64Dtype())
    winsorized_series = series.clip(lower=lower_bound, upper=upper_bound)

    return (winsorized_series)

In [ ]:
for col in kolom_numerik:
    student_mental_health[col] = teknik_winsorizing(student_mental_health[col])
    box_plot(student_mental_health[col], col, '#B07AA1')

# **Distribusi Data**

In [ ]:
import plotly.express as px
import numpy as np # Import numpy

for col in kolom_numerik:
    student_mental_health[col] = student_mental_health[col].astype(float)

for col in kolom_numerik:
    fig = px.histogram(
        student_mental_health,
        x = col,
        nbins = 25,
        color_discrete_sequence = ['#B07AA1'],
        marginal = "box",
        hover_data = student_mental_health.columns
    )

    fig.update_yaxes(
        showgrid = False,
        showticklabels=False,
        title =''
    )

    fig.update_layout(
        title={
            'text' : f'Distribusi <b><span style="color:#B07AA1"></span> {col}</b>',
            'y':0.92,
            'x':0.5,
            'xanchor': 'center',
            'yanchor': 'top'
        },
        plot_bgcolor = 'rgba(0,0,0,0)',
        bargap = 0.01,
        title_font = dict(size = 25)
    )

    fig.show()

# **Analisa Multivariate**

# **Korelasi**

In [ ]:
# Hitung korelasi
data_correlation = student_mental_health[kolom_numerik].corr()

# Atur tampilan float menjadi 2 angka desimal
pd.set_option('display.float_format', '{:.2f}'.format)

# Tampilkan hasilnya
display(data_correlation)

Age  Screen Time (hrs/day)  \
Age                           1.00                   0.01   
Screen Time (hrs/day)         0.01                   1.00   
Sleep Duration (hrs)          0.00                   0.01   
Physical Activity (hrs/week) -0.01                   0.00   

                              Sleep Duration (hrs)  \
Age                                           0.00   
Screen Time (hrs/day)                         0.01   
Sleep Duration (hrs)                          1.00   
Physical Activity (hrs/week)                 -0.03   

                              Physical Activity (hrs/week)  
Age                                                  -0.01  
Screen Time (hrs/day)                                 0.00  
Sleep Duration (hrs)                                 -0.03  
Physical Activity (hrs/week)                          1.00

In [ ]:
import plotly.express as px

# Buat Grafiknya
fig = px.imshow(
    data_correlation,
    color_continuous_scale='blues',
    title = '<b>Korelasi Kolom Numerik Data Student</b><br>',
    text_auto = True
)

#Menyembunyikan skala/rentang korelasi
fig.update_coloraxes(showscale=False)

#Atur judul heatmap
fig.update_layout(
    title = dict(
        x=0.5,
        y=0.9,
        xanchor='center',
        yanchor='top'
    ),
    width = 1000,
    height = 800
)

#Menampilkan heatmap
fig.show()

# **Statistik tiap Kategori Academic Performance Change**

In [ ]:
student_mental_health.groupby(['Academic Performance Change']).agg(
    total_data = ('Name', 'count'),
    min_age = ('Age', 'min'),
    median_age = ('Age', 'median'),
    max_age = ('Age', 'max'),
    min_screen_time = ('Screen Time (hrs/day)', 'min'),
    median_screen_time = ('Screen Time (hrs/day)', 'median'),
    max_screen_time = ('Screen Time (hrs/day)', 'max'),
    min_sleep_duration = ('Sleep Duration (hrs)', 'min'),
    median_sleep_duration = ('Sleep Duration (hrs)', 'median'),
    max_sleep_duration = ('Sleep Duration (hrs)', 'max')
)

total_data  min_age  median_age  max_age  \
Academic Performance Change                                             
Declined                            370     8.00       20.00    32.00   
Improved                            374     8.00       21.00    32.00   
Same                                481     8.00       20.00    32.00   

                             min_screen_time  median_screen_time  \
Academic Performance Change                                        
Declined                                0.00                6.80   
Improved                                0.00                7.10   
Same                                    0.00                6.70   

                             max_screen_time  min_sleep_duration  \
Academic Performance Change                                        
Declined                               17.15                1.05   
Improved                               17.15                1.05   
Same                                   17.15                1.05   

                             median_sleep_duration  max_sleep_duration  
Academic Performance Change                                             
Declined                                      6.30               11.85  
Improved                                      6.60               11.85  
Same                                          6.60               11.85

# **Data Preprocessing**

## **Feature Selection**

In [ ]:
# Copy DataFrame
data_preprocessing = student_mental_health.copy()

# Hapus kolom name
data_preprocessing = data_preprocessing.drop('Name', axis=1)

# Tampilkan hasilnya
data_preprocessing.head()

Gender   Age Education Level  Screen Time (hrs/day)  Sleep Duration (hrs)  \
0    Male 15.00         Class 8                   7.10                  8.90   
1  Female 25.00             MSc                   3.30                  1.05   
2    Male 20.00           BTech                   9.50                 11.85   
3    Male 20.00              BA                  10.80                  5.60   
4  Female 17.00        Class 11                   2.80                  5.40   

   Physical Activity (hrs/week) Stress Level Anxious Before Exams  \
0                          9.30       Medium                   No   
1                          0.20       Medium                   No   
2                          6.20       Medium                   No   
3                          5.50         High                  Yes   
4                          3.10       Medium                  Yes   

  Academic Performance Change  
0                        Same  
1                        Same  
2                        Same  
3                        Same  
4                        Same

##**Encoding**

In [ ]:
# Definisikan ulang kolom kategorik dan kolom numerik
kolom_kategorik = data_preprocessing.select_dtypes(include=['object']).columns
kolom_numerik = data_preprocessing.select_dtypes(include=['number']).columns

In [ ]:
# Import library yang dibutuhkan
from sklearn.preprocessing import OrdinalEncoder

data_preprocessing = student_mental_health.copy()
data_preprocessing = data_preprocessing.drop('Name', axis=1)

# Definisikan kolom kategorik secara eksplisit
categorical_cols_to_encode = [
    'Gender',
    'Education Level',
    'Stress Level',
    'Anxious Before Exams',
    'Academic Performance Change'
]

# Panggil object Ordinal Encoder
ord_enc = OrdinalEncoder()

# Pastikan kolom-kolom ini bertipe string sebelum di-encode
for col in categorical_cols_to_encode:
    data_preprocessing[col] = data_preprocessing[col].astype(str)

# Terapkan encoding
data_preprocessing[categorical_cols_to_encode] = ord_enc.fit_transform(data_preprocessing[categorical_cols_to_encode])

# Tampilkan hasilnya
display(data_preprocessing)

Gender   Age  Education Level  Screen Time (hrs/day)  \
0       1.00 15.00             6.00                   7.10   
1       0.00 25.00             9.00                   3.30   
2       1.00 20.00             2.00                   9.50   
3       1.00 20.00             0.00                  10.80   
4       0.00 17.00             4.00                   2.80   
...      ...   ...              ...                    ...   
1245    1.00 21.00             2.00                   6.10   
1246    1.00 15.00             7.00                   5.50   
1247    1.00 22.00            11.00                   5.50   
1248    0.00 16.00             3.00                  10.40   
1249    1.00 24.00            10.00                   2.00   

      Sleep Duration (hrs)  Physical Activity (hrs/week)  Stress Level  \
0                     8.90                          9.30          2.00   
1                     1.05                          0.20          2.00   
2                    11.85                          6.20          2.00   
3                     5.60                          5.50          0.00   
4                     5.40                          3.10          2.00   
...                    ...                           ...           ...   
1245                  6.20                          4.00          1.00   
1246                  8.00                          6.90          2.00   
1247                  4.40                          0.00          2.00   
1248                  8.00                          6.60          0.00   
1249                  8.00                          8.10          0.00   

      Anxious Before Exams  Academic Performance Change  
0                     0.00                         2.00  
1                     0.00                         2.00  
2                     0.00                         2.00  
3                     1.00                         2.00  
4                     1.00                         2.00  
...                    ...                          ...  
1245                  1.00                         0.00  
1246                  1.00                         2.00  
1247                  0.00                         0.00  
1248                  0.00                         0.00  
1249                  1.00                         2.00  

[1250 rows x 9 columns]

In [8]:
# kode untuk melakukan mapping
for col, categories in zip(categorical_cols_to_encode, ord_enc.categories_):
    print(f"Mapping untuk \033[91m{col}\033[0m:\n")
    for i, cat in enumerate(categories):
        print(f"  {cat} \u2192 {i}")
    print()

Mapping untuk Gender:

  Female → 0
  Male → 1
  Other → 2
  nan → 3

Mapping untuk Education Level:

  BA → 0
  BSc → 1
  BTech → 2
  Class 10 → 3
  Class 11 → 4
  Class 12 → 5
  Class 8 → 6
  Class 9 → 7
  MA → 8
  MSc → 9
  MTech → 10
  nan → 11

Mapping untuk Stress Level:

  High → 0
  Low → 1
  Medium → 2
  nan → 3

Mapping untuk Anxious Before Exams:

  No → 0
  Yes → 1
  nan → 2

Mapping untuk Academic Performance Change:

  Declined → 0
  Improved → 1
  Same → 2
  nan → 3



## **Splitting Data**

In [ ]:
# Tampilkan frekuensi kemunculan target setelah encoding
display(data_preprocessing['Academic Performance Change'].value_counts())

Academic Performance Change
2.00    486
1.00    377
0.00    372
3.00     15
Name: count, dtype: int64

📙 **Catatan**
* Declined → 0
* Improved → 1
* Same → 2

## **Imbalance Target**

In [ ]:
# Import library imblearn
from imblearn.over_sampling import SMOTE

# Pastikan tidak ada NaN di X_train atau y_train sebelum SMOTE
valid_indices = X_train.dropna().index
X_train = X_train.loc[valid_indices]
y_train = y_train.loc[valid_indices]

# object SMOTE
smote = SMOTE(random_state=42)

# Terapkan pada X_train dan y_train dan buat variabel baru hasil oversampling
X_train_resample, y_train_resample = smote.fit_resample(X_train, y_train)

In [ ]:
# Tampilkan frekuensi kemunculan target
display(y_train_resample.value_counts())

Academic Performance Change
1.00    373
0.00    373
2.00    373
3.00    373
Name: count, dtype: int64

# **Modelling**

In [ ]:
from imblearn.pipeline import Pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

pipeline = Pipeline(steps=[
    ('scaler', RobustScaler()),
    ('smote', SMOTE()),
    ('model',DecisionTreeClassifier())
])

param_grid = {
    'smote__k_neighbors': [3, 5, 7],
    'model__criterion': ['gini', 'entropy', 'log_loss'],
    'model__max_depth': [None, 5, 10, 20, 30],
    'model__min_samples_split': [2, 5, 10],
    'model__min_samples_leaf': [1, 2, 5],
    'model__class_weight': [None, 'balanced']
}

grid = GridSearchCV(
    estimator = pipeline,
    param_grid = param_grid,
    cv = 5,
    scoring = 'f1_macro',
    n_jobs = -1
)

grid.fit(X_train, y_train)

print("Best Params:", grid.best_params_)
print("Best CV Score:", grid.best_score_)

Best Params: {'model__class_weight': None, 'model__criterion': 'entropy', 'model__max_depth': 30, 'model__min_samples_leaf': 1, 'model__min_samples_split': 2, 'smote__k_neighbors': 5}
Best CV Score: 0.3671178912134264


In [ ]:
# Import library yang dibutuhkan
from sklearn.model_selection import StratifiedKFold, cross_val_score

# Masukan parameter yang telah di-tuning
model_dtc = DecisionTreeClassifier(
    criterion = 'entropy',
    max_depth = 30,
    min_samples_leaf = 1,
    min_samples_split = 5,
    splitter = 'best'
)

# Definsikan pipeline baru
new_pipe = Pipeline([
    ('scaler', RobustScaler()),
    ('smote', SMOTE()),
    ('model', model_dtc)
])

# Terapkan pada CV
cv = StratifiedKFold(
    n_splits = 5,
    shuffle = True,
    random_state = 42
)

# Pastikan tidak ada NaN di X atau y sebelum cross_val_score
valid_indices_for_cv = X.dropna().index
X_cleaned_for_cv = X.loc[valid_indices_for_cv]
y_cleaned_for_cv = y.loc[valid_indices_for_cv]

scores = cross_val_score(new_pipe, X_cleaned_for_cv, y_cleaned_for_cv, cv=cv, scoring = 'accuracy')

print(f'Scores : {scores}')
print(f'Mean   : {scores.mean()}')
print(f'Std    : {scores.std()}')

Scores : [0.46058091 0.47083333 0.43333333 0.475      0.47083333]
Mean   : 0.46211618257261405
Std    : 0.015157702000531837


In [ ]:
model_dtc.fit(X_train_resample, y_train_resample)

# Predict test set labels
y_pred = model_dtc.predict(X_test)

In [ ]:
from sklearn import metrics

# Menghitung dan mencetak laporan klasifikasi
classification_report = metrics.classification_report(y_pred, y_test)

# Tampilkan hasilnya
print(classification_report)

              precision    recall  f1-score   support

         0.0       0.44      0.45      0.45        73
         1.0       0.48      0.43      0.46        83
         2.0       0.47      0.51      0.49        90
         3.0       0.00      0.00      0.00         4

    accuracy                           0.46       250
   macro avg       0.35      0.35      0.35       250
weighted avg       0.46      0.46      0.46       250



In [ ]:
import plotly.express as px
from sklearn.metrics import confusion_matrix

# Plot dengan Plotly
fig = px.imshow(
    confusion_matrix(y_test, y_pred),
    text_auto = True,
    color_continuous_scale = 'Blues',
    title = "<b>Confusion Matrix Model Decision Tree</b>",
)

# Ubah tampilan xticks dan yticks
fig.update_xaxes(
    tickmode = "array",
    tickvals = [0, 1, 2],
    ticktext = ['Declined', 'Improved', 'Same']
)

fig.update_yaxes(
    tickmode = "array",
    tickvals = [0, 1, 2],
    ticktext = ['Declined', 'Improved', 'Same'],
    tickangle = -90
)

# Hapus legend / colorbar
fig.update_layout(coloraxis_showscale = False)

# Judul dan label + ukuran figure
fig.update_layout(
    title = dict(
        x = 0.5,
        y = 0.9,
        xanchor = 'center',
        yanchor = 'top'
    ),
    xaxis_title = "<b>Prediksi</b>",
    yaxis_title = "<b>Nilai Sebenarnya</b>",
    width = 700,
    height = 700
)

# Tampilkan grafik
fig.show()

# 💡 **Insight**
Model paling akurat dalam memprediksi kategori "Same" dengan $46$ data yang tertebak benar. Ada dua kemungkinan hal ini terjadi:
* Jumlah Data: Kategori "Same" mungkin memiliki jumlah sampel terbanyak dalam dataset, sehingga model lebih sering "berlatih" mengenali pola ini
* Karakteristik Unik: Fitur (kolom) dalam datamu mungkin memiliki pola yang sangat jelas untuk mahasiswa yang kondisi mentalnya tidak berubah, sehingga lebih mudah dibedakan dari kategori "Improved" atau "Declined".

### **REKOMENDASI**


1.   Keseimbangan Layar & Tidur

Pada data terlihat bahwa siswa yang memiliki waktu tidur kurang dari 7 jam atau penggunaan layar (screen time) lebih dari 6 jam cenderung memiliki tingkat stres yang lebih tinggi. Secara biologis, kurang tidur mengganggu fungsi otak dalam mengatur emosi dan fokus. Sementara itu, paparan cahaya biru (blue light) dari layar dapat menekan produksi melatonin, sehingga kualitas tidur semakin menurun.

2.   Aktivitas Fisik sebagai Penyangga Kebugaran

Dengan melihat konsep regulasi emosi melalui aktivitas fisik. Secara akademis, olahraga meningkatkan plastisitas otak dan menurunkan kortisol (hormon stres). Korelasi negatif antara frekuensi olahraga dan tingkat stres sejalan dengan korelasi mahasiswa yang kurang bergerak cenderung memiliki akumulasi energi negatif yang tidak tersalurkan.



<br>
<a href="https://www.linkedin.com/in/ishmahazatilsyadzwina/"><img src="https://img.shields.io/badge/-© 2025 Ishmah Azatil Syadzwina-417DAC?style=for-the-badge&logoColor=white"/></a>

<a href="https://dqlab.id/"><img src="https://raw.githubusercontent.com/bachtiyarma/Material/main/Image/Materi-Python/FINAL LOGO GSB 2025 - ADV (Buat IG, TikTok, Podcast) (2).png" align="left" width="15%" /></a>